src/Administration/Resources/app/administration/src/app/asyncComponent/form/sw-datepicker/index.js

```diff
@@ -388,6 +384,10 @@ export default {
          * Create the flatpickr instance. If already one exists it will be recreated.
          */
         createFlatpickrInstance() {
+            if (!this.$refs.flatpickrInput) {
```

Comment:
How will this help other than preventing errors :thinking: 

1. Component gets loaded without the ref being loaded, which is caused by the component being async
2. Early return is triggered
3. Rerender would only be triggered if:
4. The component config changes
5. The dateType changes
6. The locale changes

According to your description and the code the component should not be usable after reloading the page.
Can you please elaborate on this and write a test case proving that it works?

Reply:
This would be correct, if loading the async component would not trigger remounts of the underlying root component, but it does, as stated (or as i now realize, more "hinted") in the description of this PR:
> The fact that multiple errors (and not a single one) occur is likely connected to the sw-datepicker being async, but this is not further investigated there.

This `if` statement will only have an effect if the component is mounted at a time when it cannot be mounted. If a further fix for async components to not remount the base component multiple times resolves to only mount them when able to be mounted, this will have no influence on the functionality of the component.

I will try to add a test for this, but can not guarantee results considering the "one-off per vue runtime" nature of the bug.

Reply:
I was not able to create a test, however i understand the root cause better now. In Vue 2, [as stated here in the docs](https://v2.vuejs.org/v2/api/#mounted), Vue calling the `mounted` event of a component does not mean all its subcomponents are mounted yet. As in the component definition of the underlying [`sw-contextual-field`](https://github.com/shopware/platform/blob/84a4e9a0d4269451ee94e41bc8bcf0d068ff16c5/src/Administration/Resources/app/administration/src/app/component/form/field-base/sw-contextual-field/sw-contextual-field.html.twig#L6) and its underlying [`sw-block-field`](https://github.com/shopware/platform/blob/717e5952ad6f5ef4d5a30ce30e3f6a3c7e789af7/src/Administration/Resources/app/administration/src/app/component/form/field-base/sw-block-field/sw-block-field.html.twig#L7) all child events are bubbled up the component graph, we get the following hooks in the following order:
1. The `hook:mounted` from `sw-contextual-field`
2. The `hook:mounted` from `sw-block-field`
3. The `hook:mounted` from `sw-base-field`

Only when the last child is mounted, the inner `input` is also mounted and available as a `$ref`, which means only on the last `hook:mounted` we can create a flatpickr instance.

This should [not be an issue anymore in Vue3](https://vuejs.org/api/options-lifecycle.html#mounted), as when a `hook:mounted` is called it is ensured that all *synchronous* child components (note that all components in the form field hierarchy are of this type) are already also mounted. This would not resolve the issue with multiple calls, now however each call will be able to build a flatpickr instance as we receive them in the inverted above order.

I think the reason i cannot replicate this in a test is that the underlying Vue instance is different for `vue-test-utils`.

Reply:
Currently working on Vue 3 I unfortunately have to tell you it's even more of a problem there :see_no_evil: In Vue 3 the `AsyncComponentWrapper` also emits mounted hooks once mounted :see_no_evil: Therefore every async component expecting `$refs` to be available in created or mounted may fail. 

My suggestion:

1. I think the safest way to go about this, is to either emit an event from the reffed component and wait for that event.
2. Watch `this.$refs` for changes and once the ref is there run the initialization

What do you think could this solve the issues?

Reply:
Let me answer the two suggestions separately:
1. As the reffed component is not a real Vue component but merely an HTML `input` element, it does not emit Vue hooks and has no "mounted" state. If it exists, it is always mounted. In theory we could workaround this by wrapping it in a simple Vue component which is merely more than a `div` with a slot (which would then be `sw-native-element-wrapper`) or even the `input` itself (which would then be `sw-native-input-wrapper`). This new component is then able to emit events and thus signal that the input is mounted.
2. As stated in the PR (under section 2) and[ here in the docs](https://v2.vuejs.org/v2/guide/components-edge-cases.html#Accessing-Child-Component-Instances-amp-Child-Elements) (scroll down a bit to the red box), `$refs` are not reactive, thus we cannot watch them for changes (which is the reason the computed property was not working in the first place).

Let me know what you think, in the meantime i will try 1. and see if it even works.

Reply:
Ah now I got you :+1: 

Regarding 1. the input is already part of a component and that should be the base field. We could simply emit a `base-field-mounted` in the onMounted method. That way we can filter in any component that the input element is truly mounted.

Reply:
I will give this a shot :)

Reply:
This seems to work just fine. I have pushed the changes.